forte/data/extractor/extractor.py

```diff
@@ -0,0 +1,230 @@
+
+
+class CharExtractor(BaseExtractor):
+    def __init__(self, config: Dict):
+        self.max_char_length = getattr(config, "max_char_length", None)
+        super().__init__(config)
+
+    def update_vocab(self, pack: DataPack, instance: EntryType):
+        for word in pack.get(self.entry, instance):
+            for char in word.text:
+                self.add_entry(char)
+
+    def extract(self, pack: DataPack, instance: EntryType):
+        data = []
+        max_char_length = -1
+
+        for word in pack.get(self.entry, instance):
+            tmp = []
+            for char in word.text:
+                tmp.append(self.entry2id(char))
+            data.append(tmp)
+            max_char_length = max(max_char_length, len(tmp))
+
+        if self.max_char_length is not None:
+            max_char_length = min(self.max_char_length, max_char_length)
+
+        return Feature(data, self.get_pad_id(), 2)
+
+
+class AnnotationSeqExtractor(BaseExtractor):
+    def __init__(self, config: Dict):
+        super().__init__(config)
+        self.entry = config["entry"]
+        self.attribute = config["attribute"]
+        self.strategy = config["strategy"]
+        self.based_on = config["based_on"]
+
+    @classmethod
+    def bio_variance(cls, tag):
+        return [(tag, "B"), (tag, "I"), (None, "O")]
+
+    @classmethod
+    def bio_tag(cls, instance_based_on, instance_entry):
+        tagged = []
+        cur_entry_id = 0
+        prev_entry_id = None
+        cur_based_on_id = 0
+
+        while cur_based_on_id < len(instance_based_on):
+            base_begin = instance_based_on[cur_based_on_id].begin
+            base_end = instance_based_on[cur_based_on_id].end
+
+            if cur_entry_id < len(instance_entry):
+                entry_begin = instance_entry[cur_entry_id].begin
+                entry_end = instance_entry[cur_entry_id].end
+            else:
+                lastone = len(instance_based_on) - 1
+                entry_begin = instance_based_on[lastone].end + 1
+                entry_end = instance_based_on[lastone].end + 1
+
+            if base_end < entry_begin:
+                # Base: [...]
+                # Entry       [....]
+                tagged.append((None, "O"))
+                prev_entry_id = None
+                cur_based_on_id += 1
+            elif base_begin >= entry_begin and base_end <= entry_end:
+                # Base:    [...]
+                # Entry:  [.......]
+                if prev_entry_id == cur_entry_id:
+                    tagged.append((instance_entry[cur_entry_id], "I"))
+                else:
+                    tagged.append((instance_entry[cur_entry_id], "B"))
+                prev_entry_id = cur_entry_id
+                cur_based_on_id += 1
+            elif base_begin > entry_end:
+                # Base:         [...]
+                # Entry: [....]
+                cur_entry_id += 1
+            else:
+                raise AssertionError("Unconsidered case.")
```

Comment:
A warning here might be better. I think that partial overlap is not impossible, especially when the tokenization and ner comes from two sources. We probably can include the partially overlapped token and give it a tag. And we should also document this behavior clearly.

Reply:
Got it.